### PR TITLE
Update helm to v3.8.1 + Add OCI Support

### LINKED ISSUE #633: Add registry support (OCI)
Seems helm3 OCI registry is working properly. I would be nice to be able to deploy from docker repos. 
I might be missing something,  I tried some combinations with repo, but nothing wokred out. So - I expect this is probably new feature.

I got idea, that it is experimental feature, but as meny experimental features, which work  - it will stay forever.

### LINKED ISSUE #666: Support for oci:// scheme (Helm 3.5.0)
### Description

Helm 3.5.0 now supports `oci://` scheme for pulling dependencies from OCI registries. Can we upgrade to it?

### Potential Terraform Configuration

```hcl
resource "helm_release" "prometheus_operator" {
  name       = "prometheus-operator"
  chart      = "kube-prometheus-stack"
  version    = "12.11.3"
  repository = "oci://..."
  namespace  = "default"
}
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### LINKED ISSUE #827: Use helm 3.8.0
### Description

Currently the provider is using helm 3.7.2, but helm 3.8.0 is now released - currently we're having to prevent upgrading our local helm version due to mismatch in how these two versions of helm operate when it comes to pulling OCI charts from private registries. (we're having to use null resources to login to private registries to pull OCI charts).

It would be good if the provider is updated to use the latest version of helm as well.

### PR DESCRIPTION
**Pull request by jrhouston**
### Description

This PR updates the helm dependency to the latest version.

Closes #827 
Closes #666 
Closes #633 

### Acceptance tests
- [x] Have you added an acceptance test for the functionality being added?

### Release Note
Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-helm/blob/main/CHANGELOG.md):

```release-note
* Update helm to v3.8.1
* Add support for OCI registries
```
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### CODE REVIEW

**Comment by sfro** [comment]
I'm kind of new to terraform-helm-provider, so I'm drawing a blank on how I can get it to create the RegistryClient using this branch. Is it something I need to set in the helm_release resource in my terraform file?

I made an ugly hack here that can pull from OCI registries: https://github.com/sfro/terraform-provider-helm/pull/1

**Comment by jrhouston** [comment]
Thanks for investigating that @sfro – I was puzzling over how to deal with the `registryClient` being private in `ChartPathOptions`. I think what I'm going to do is update this PR to have the `chartPathOptions()` accept the `client` and update the existing `ChartPathOptions` rather than create a new instance of it, like you've done in the branch you linked.

**Comment by sfro** [comment]
Ah, that makes sense! Better than explicitly copying over all the exported fields like I did.

**Comment by LP0101** [comment]
I built this branch to test OCI repos, however I'm still getting an error when trying to install them:

```
│ Error: unable to lookup chart "oci://myregistry.azurecr.io/helm/mychart", missing registry client
│ 
│   with helm_release.test,
│   on main.tf line 32, in resource "helm_release" "test":
│   32: resource "helm_release" "test" {
```

With the following manifest:

```
resource "helm_release" "test" {

  name = "test"
  namespace = "luca-test" 
  chart = "oci://myregistry.azurecr.io/helm/mychart"
  version = "6.9.3"

  repository_username = "foo"
  repository_password = "bar"
}
```

File: helm/provider.go [hunk 1]
```diff
@@ -373,10 +375,52 @@ func (m *Meta) GetHelmConfiguration(namespace string) (*action.Configuration, er
...
+	if err != nil {
+		return fmt.Errorf("could not create OCI registry client: %v", err)
+	}
+	actionConfig.RegistryClient = registryClient
+
+	// log in to the registry if neccessary
+	repository := d.Get("repository").(string)
+	chartName := d.Get("name").(string)
+	var ociURL string
+	if registry.IsOCI(repository) {
```
Comment (alexsomesan) [inline helm/provider.go:400]: Is there an opportunity here for users to set both `name` and `repository`?
Do we validate against that?

File: helm/provider.go [hunk 2]
```diff
@@ -373,10 +375,52 @@ func (m *Meta) GetHelmConfiguration(namespace string) (*action.Configuration, er
...
+		ociURL = repository
+	} else if registry.IsOCI(chartName) {
+		ociURL = chartName
+	}
+	if ociURL != "" {
```
Comment (alexsomesan) [inline helm/provider.go]: Why not bail out early on the inverse condition here?

File: helm/provider.go [hunk 3]
```diff
@@ -373,10 +375,52 @@ func (m *Meta) GetHelmConfiguration(namespace string) (*action.Configuration, er
...
+		username := d.Get("repository_username").(string)
+		password := d.Get("repository_password").(string)
+		if username != "" && password != "" {
```
Comment (alexsomesan) [inline helm/provider.go]: Should we log an error here that the credentials were invalid / insufficient?

**Review by alexsomesan — COMMENTED** [review]
Looks good. Wil try it out locally next.

Left some questions below about the potential combinations of inputs and how to validate them. Mostly to clarify my understanding about how this is supposed to be used.

File: helm/provider.go (hunk 1 above)
Comment (jrhouston) [reply helm/provider.go:400]: Yes they can either use the full oci URL to the chart as the `name`, or use the oci registry URL as the `repository` and set just the `name` to the name of the chart in the registry. Not sure what kind of additional validation we would do there. 

edit: to be clear if you set both repository AND name only repository can be an OCI registry – that's why it's a else if statement blow.

File: helm/provider.go (hunk 2 above)
Comment (jrhouston) [reply helm/provider.go]: Good shout I'll shuffle it around.

File: helm/provider.go (hunk 3 above)
Comment (jrhouston) [reply helm/provider.go]: The intent here is just to check if a username and password has been set, and if so to perform the login operation. The login operation isn't actually needed unless the registry requires a username and password, so the user will get an error if they haven't specified it anyway.

File: helm/provider.go [hunk 4]
```diff
@@ -373,10 +375,55 @@ func (m *Meta) GetHelmConfiguration(namespace string) (*action.Configuration, er
...
+func OCIRegistryLogin(actionConfig *action.Configuration, d dataGetter) error {
+	registryClient, err := registry.NewClient()
+	if err != nil {
+		return fmt.Errorf("could not create OCI registry client: %v", err)
+	}
+	actionConfig.RegistryClient = registryClient
+
+	// log in to the registry if neccessary
+	repository := d.Get("repository").(string)
+	chartName := d.Get("name").(string)
```
Comment (arybolovlev) [inline helm/provider.go]: Shall we use here `chart`(chart name) instead of `name`(release name)?

Comment (jrhouston) [reply helm/provider.go]: Yep you're right, good catch

**Review by arybolovlev — APPROVED** [review]
Great, @jrhouston! 🚀

**Comment by joscabmar** [comment]
Are there any plans to tag a new release so that we can start using this feature with the provider? I saw the last tagged release is from November 2021.

**Comment by jrhouston** [comment]
Yes, I'll be cutting a release shortly @joscabmar

**Comment by joscabmar** [comment]
> Yes, I'll be cutting a release shortly @joscabmar

That is great! Thanks a lot!

**Pull request merged**